src/NHibernate/Type/GenericIdentifierBagType.cs

```diff
@@ -82,5 +84,71 @@ protected override void Add(object collection, object element)
 		{
 			((IList<T>)collection).Add((T)element);
 		}
+
+		protected override bool AreCollectionElementsEqual(IEnumerable original, IEnumerable target)
+		{
+			var first = (IList<T>)original;
+			var second = (IList<T>)target;
+
+			return first.Count == second.Count &&
+			       first.All(element => first.Count(x => x.Equals(element)) == second.Count(x => x.Equals(element)));
+		}
+
+		public override object ReplaceElements(
+			object original,
+			object target,
+			object owner,
+			IDictionary copyCache,
+			ISessionImplementor session)
+		{
+			var elemType = GetElementType(session.Factory);
+			var targetPc = target as PersistentIdentifierBag<T>;
+			var originalPc = original as IPersistentCollection;
+			var iter = (IList<T>)original;
+			var clearTargetsDirtyFlag = false;
+			var clearTarget = true;
+
+			if (targetPc != null)
+			{
+				if (originalPc == null)
+				{
+					if (!targetPc.IsDirty && AreCollectionElementsEqual(iter, (IEnumerable)target))
+					{
+						clearTargetsDirtyFlag = true;
+						clearTarget = false;
+					}
+				}
+				else
+				{
+					if (!originalPc.IsDirty)
+					{
+						clearTargetsDirtyFlag = true;
+					}
+				}
+			}
+
+			if (clearTarget)
+			{
+				Clear(target);
+				foreach (var obj in iter)
+				{
+					Add(target, elemType.Replace(obj, null, session, owner, copyCache));
+				}
+			}
+			else
+			{
+				for (var i = 0; i < targetPc.Count; i++)
+				{
+					targetPc[i] = (T)elemType.Replace(iter[i], null, session, owner, copyCache);
```

Comment:
`iter` (to be renamed `originalIter` maybe) can be equal without having the same ordering of elements (order is not significant in bags). `iter[i]` can be another element than `targetPc[i]`. We need to search in `iter` the element corresponding to `targetPc[i]`. Otherwise `targetPc` may swap its internal identifiers between elements, which can result in deleting the wrong element in db if we remove one element from it later on.